edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/SystemInfo.cs

```diff
@@ -4,44 +4,69 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Core
     using System.Text;
 
     using static System.Net.WebUtility;
 
     public class SystemInfo
     {
-        public SystemInfo(string operatingSystemType, string architecture, string version, IDictionary<string, object> additionalProperties)
+        public SystemInfo(string operatingSystemType, string architecture, string version, ProvisioningInfo provisioning, string serverVersion, string kernelVersion, string operatingSystem, int numCpus, string virtualized, IReadOnlyDictionary<string, object> additionalProperties)
         {
             this.OperatingSystemType = operatingSystemType;
             this.Architecture = architecture;
             this.Version = version;
+            this.Provisioning = provisioning;
+            this.ServerVersion = serverVersion;
+            this.KernelVersion = kernelVersion;
+            this.OperatingSystem = operatingSystem;
+            this.NumCpus = numCpus;
+            this.Virtualized = virtualized;
             this.AdditionalProperties = additionalProperties;
         }
 
+        public SystemInfo(string operatingSystemType, string architecture, string version, IReadOnlyDictionary<string, object> additionalProperties)
+            : this(operatingSystemType, architecture, version, ProvisioningInfo.Empty, string.Empty, string.Empty, string.Empty, 0, string.Empty, additionalProperties)
+        {
+        }
+
         public SystemInfo(string operatingSystemType, string architecture, string version)
             : this(operatingSystemType, architecture, version, new Dictionary<string, object>())
         {
         }
 
-        // NOTE: changed to snake-case to reflect edgelet
         public string OperatingSystemType { get; }
 
         public string Architecture { get; }
 
         public string Version { get; }
 
-        // NOTE: changed to IDictionary from IReadOnlyDictionary since the
-        // latter cannot be used as extension data.  Likewise for <string,
-        // object> from <string, string>.
-        [Newtonsoft.Json.JsonExtensionData]
-        public IDictionary<string, object> AdditionalProperties { get; }
+        public ProvisioningInfo Provisioning { get; }
+
+        public string ServerVersion { get; }
+
+        public string KernelVersion { get; }
+
+        public string OperatingSystem { get; }
+
+        public int NumCpus { get; }
+
+        public string Virtualized { get; }
+
+        [Newtonsoft.Json.JsonIgnore]
+        public IReadOnlyDictionary<string, object> AdditionalProperties { get; }
 
         public string ToQueryString()
         {
-            // NOTE (from author): Reflection will not work due to name mappings
             StringBuilder b = new StringBuilder()
-                .Append($"kernel_name={UrlEncode(this.OperatingSystemType ?? string.Empty)};")
-                .Append($"cpu_architecture={UrlEncode(this.Architecture ?? string.Empty)};");
+                .Append($"kernel={UrlEncode(this.OperatingSystemType ?? string.Empty)};")
```

Comment:
Does this control how the metric is serialized, and the attributes named/cased? 

For instance, the in https://docs.microsoft.com/en-us/azure/iot-edge/how-to-access-built-in-metrics?view=iotedge-2020-11#available-metrics, the attribute is named `Architecture`, where-as in this code, the attribute is named `architecture`

Similarly the link states `OperatingSystemType` for the OS, but the code below has `operating_system`

Reply:
The casing is intentional. A previous iteration used the reflection APIs and serialized the field names verbatim, but since internal dashboards are already using the snake-case naming, it would be easier for everyone to just explicitly synthesize the query string.

`OperatingSystemType` actually corresponds most closely to the kernel name instead of the operating system name: [ref](https://docs.docker.com/engine/api/v1.41/#operation/SystemInfo).